src/torchmetrics/functional/regression/utils.py

```diff
@@ -28,3 +28,14 @@ def _check_data_shape_to_num_outputs(preds: Tensor, target: Tensor, num_outputs:
             f"Expected argument `num_outputs` to match the second dimension of input, but got {num_outputs}"
             f" and {preds.shape[1]}."
         )
+
+
+def _check_data_shape_to_weights(preds: Tensor, weights: Tensor) -> None:
+    """Check that the predictions (or target) and weights have the correct shape, else raise error."""
+    if weights.ndim == preds.ndim:
+        raise ValueError(
+            f"Expected `n_samples` of preds to match the length of `weights`,"
+            f" but got {preds.shape[0]} and {len(weights)}."
+        )
+    else:
+        raise ValueError("")
```

Comment:
```suggestion
    if weights.shape != preds.shape:
        raise ValueError(
            f"Expected `n_samples` of preds to match the length of `weights`,"
            f" but got {preds.shape[0]} and {len(weights)}."
        )
```
I think this check needs to be even more restrictive such that they can be multiplied together or else I am missing something.
Also I do not think there should be an `else` statement.

Reply:
I split the check into single and multi-output cases:
- if single output (`preds.ndim=1`), `preds` and `weights` must have same shape
- if multi-output (`preds.ndim=2`)
    - if `weights.ndim=2`, `preds` and `weights` must have same shape
    - if `weight.ndim=1`, `preds.shape[0]` has to be equal to `len(weights)`

Reply:
Okay, that makes sense to me, but the implementation then needs to reflect that. Currently, if `weight.ndim=1` and `preds.ndim=2` the code will just fail with an value error.